vllm/attention/backends/flash_attn.py

```diff
@@ -612,80 +816,104 @@ def unified_flash_attention(
+            # not cached. This happens during the initial memory profiling run.
+            torch.ops._C_cache_ops.reshape_and_cache_flash(
+                key,
+                value,
+                kv_cache[0],
+                kv_cache[1],
+                updated_slot_mapping.flatten()
+                if updated_slot_mapping is not None else None,
+                kv_cache_dtype,
+                k_scale,
+                v_scale,
+            )
 
-        # Reshape the input keys and values and store them in the cache.
-        # If kv_cache is not provided, the new key and value tensors are
-        # not cached. This happens during the initial memory profiling run.
-        torch.ops._C_cache_ops.reshape_and_cache_flash(
-            key,
-            value,
-            kv_cache[0],
-            kv_cache[1],
-            attn_metadata.slot_mapping.flatten(),
-            kv_cache_dtype,
-            k_scale,
-            v_scale,
-        )
-
-    num_prefill_tokens = attn_metadata.num_prefill_tokens
-    num_decode_tokens = attn_metadata.num_decode_tokens
-    assert key.shape[0] == num_prefill_tokens + num_decode_tokens, \
-                f"key : {key.shape} : #prefill tokens {num_prefill_tokens} : #decode tokens {num_decode_tokens}" # noqa
-    assert value.shape[0] == num_prefill_tokens + num_decode_tokens, \
-                f"value : {value.shape} : #prefill toks {num_prefill_tokens} : #decode toks {num_decode_tokens}" # noqa
-
-    # Query for decode. KV is not needed because it is already cached.
+    num_prefill_tokens, num_encoder_tokens,  num_decode_tokens = \
+        _get_num_prefill_encode_decode_tokens(attn_metadata, attn_type)
     decode_query = query[num_prefill_tokens:]
     # QKV for prefill.
     query = query[:num_prefill_tokens]
-    key = key[:num_prefill_tokens]
-    value = value[:num_prefill_tokens]
-
     assert query.shape[0] == num_prefill_tokens
     assert decode_query.shape[0] == num_decode_tokens
 
     prefill_output: Optional[torch.Tensor] = None
     decode_output: Optional[torch.Tensor] = None
-
     if prefill_meta := attn_metadata.prefill_metadata:
         # Prompt run.
         if (kv_cache.numel() == 0 or prefill_meta.block_tables is None
                 or prefill_meta.block_tables.numel() == 0):
             # normal attention
             # When block_tables are not filled, it means q and k are the
             # prompt, and they have the same length.
+            q_seq_start_loc, q_seq_len, k_seq_start_loc, k_seq_len = \
+                _get_query_key_seq_metadata(prefill_meta, True, attn_type)
+
+            if (attn_type == AttentionType.ENCODER or \
+                attn_type == AttentionType.ENCODER_DECODER):
+                key = key[:num_encoder_tokens]
+                value = value[:num_encoder_tokens]
+            else:
+                key = key[:num_prefill_tokens]
+                value = value[:num_prefill_tokens]
```

Comment:
If you set `def _get_num_prefill_encode_decode_tokens() if attn_type == AttentionType.DECODER: num_encoder_tokens=attn_metadata.num_prefill_tokens` like what xformer is doing, you can avoid this branch. The `num_encoder_token` is similar to `q_len`. 

And not sure if it is correct to remove these lines and pass the full key and value to the attention kernel.

Reply:
I am slightly in favor of this because when the AttentionType is DECODER I was thinking it might be more intuitive to split the key based on prefill_tokens rather than encoder_tokens. encoder_tokens seem more relevant when the AttentionType is ENCODER or ENCODER_DECODER. 

I modified the xformers code also to have the same split since I am now using the common get_num_prefill_encode_decode_tokens. Please let me know your preference and I will update this accordingly in both the backends.

Reply:
I think the goal of `_get_num_prefill_encode_decode_tokens` is to unify different attention types and avoid branching in the following code path as much as possible. What about renaming the three variables to make it clearer, e.g., num_prefill_query_tokens, num_prefill_kv_tokens, num_decode_query_tokens?

Reply:
Done renamed to num_prefill_query_tokens, num_prefill_kv_tokens, num_decode_query_tokens and removed the if branches in the backends.